src/trace-plugin-interface.js

```diff
@@ -0,0 +1,242 @@
+'use strict';
+var cls = require('./cls.js');
+var constants = require('./constants.js');
+
+/**
+ * This file describes an interface for third-party plugins to enable tracing
+ * for arbitrary modules.
+ */
+
+/**
+ * An object that is associated with a single child span. It exposes
+ * functions for adding labels to or closing the associated span.
+ */
+function ChildSpan(agent, context) {
+  this.agent = agent;
+  this.context = context;
+}
+
+/**
+ * Adds a label to the underlying span.
+ * @param {string} key The name of the label to add.
+ * @param {*} value The value of the label to add.
+ */
+ChildSpan.prototype.addLabel = function(key, value) {
+  this.context.addLabel(key, value);
+};
+
+/**
+ * Ends the underlying span. This function should only be called once.
+ */
+ChildSpan.prototype.endSpan = function() {
+  this.context.close();
+};
+
+/**
+ * An object that is associated with a single root span. It exposes
+ * functions for adding labels to or closing the associated span.
+ */
+function Transaction(agent, context) {
+  this.agent = agent;
+  this.context = context;
+  this.closed = false;
+}
+
+/**
+ * Adds a label to the underlying span.
+ * @param {string} key The name of the label to add.
+ * @param {*} value The value of the label to add.
+ */
+Transaction.prototype.addLabel = function(key, value) {
+  this.context.addLabel(key, value);
+};
+
+/**
+ * Ends the underlying span. This function should only be called once.
+ */
+Transaction.prototype.endSpan = function() {
+  this.context.close();
+};
+
+/**
+ * Runs the given function in a child span, passing it an object that
+ * exposes an interface for adding labels and closing the span.
+ * @param {object} options An object that specifies options for how the child
+ * span is created and propogated.
+ * @param {string} options.name The name to apply to the child span.
+ * @param {?function(string, string)} options.setHeader A function describing
```

Comment:
After discussing the role of the trace context header with Matt I'm now convinced that it's better to generalize the functionality of `setHeader` to `setTraceContext` since it's not necessarily true that trace context is only propagated over the network through HTTP headers. See 2629880.

I think it'd be best to discuss the importance of providing this optional parameter in the plugin implementation guide rather than the JSDoc.